File: src/main/java/org/noxet/noxetserver/util/InventoryCoordinate.java
package org.noxet.noxetserver.util;

public interface InventoryCoordinate {
    int getX();
    int getY();

    default int getSlotIndex() {
        return getY() * 9 + getX();
    }

    default boolean isAt(InventoryCoordinate coordinate) {
        return getSlotIndex() == coordinate.getSlotIndex();
    }
}

File: src/main/java/org/noxet/noxetserver/util/InventoryCoordinate.scala
package org.noxet.noxetserver.util

case class InventoryCoordinate(x: Int, y: Int):
    def slotIndex: Int = y * 9 + x

object InventoryCoordinate:
    given Conversion[(Int, Int), InventoryCoordinate] = (x, y) => InventoryCoordinate(x, y)
    given Conversion[Int, InventoryCoordinate] = (slotIndex) => InventoryCoordinate(slotIndex)

    def apply(slotIndex: Int): InventoryCoordinate =
        InventoryCoordinate(slotIndex % 9, slotIndex / 9)